Assets/_BForBoss/_Weapons/Scripts/Weapons/WeaponsManager.cs

```diff
@@ -0,0 +1,131 @@
+        [Tooltip("How fast the weapon goes back to it's original position after the recoil is finished")]
+        [SerializeField]
+        private float _recoilRestitutionSharpness = 10f;
+
+        private float _weaponBobFactor = 0;
+        private Vector3 _weaponBobLocalPosition;
+        private Vector3 _accumulatedRecoil;
+        private Vector3 _weaponRecoilLocalPosition;
+
+        public void Initialize(ICharacterMovement characterMovement)
+        {
+            _characterMovement = characterMovement;
+        }
+
+        private void Start()
+        {
+            _equipmentBehaviour.Weapons.ForEach(weapon => weapon.OnFireWeapon += OnWeaponFired);
+        }
+
+        private void OnDestroy()
+        {
+            _equipmentBehaviour.Weapons.ForEach(weapon => weapon.OnFireWeapon -= OnWeaponFired);
+        }
+
+        private void OnWeaponFired(int _)
+        {
+            _accumulatedRecoil += Vector3.back * _equipmentBehaviour.CurrentWeapon.VisualRecoilForce;
+            _accumulatedRecoil = Vector3.ClampMagnitude(_accumulatedRecoil, _maxRecoilDistance);
+        }
+
+        private void OnValidate()
+        {
+            if (_equipmentBehaviour == null)
+            {
+                Debug.LogWarning("Equipment Behaviour missing from WeaponsManager");
+            }
+        }
+
+        private void LateUpdate()
+        {
+            UpdateBobbingWeaponOnMovement();
+            UpdateWeaponRecoil();
+
+            _weaponHolder.transform.localPosition = _weaponBobLocalPosition + _weaponRecoilLocalPosition;
+        }
+
+        private void UpdateBobbingWeaponOnMovement()
+        {
+            var characterVelocity = _characterMovement.CharacterVelocity;
+            var characterMovementFactor = 0f;
+            if (CanBobWeapon())
+            {
+                characterMovementFactor =
+                    Mathf.Clamp01(characterVelocity.magnitude / _characterMovement.CharacterMaxSpeed);
+            }
+            
+            _weaponBobFactor = Mathf.Lerp(_weaponBobFactor, characterMovementFactor, _weaponBobSharpness * Time.deltaTime);
+            
+            var hBobValue = Mathf.Sin(Time.time * _weaponBobFrequency) * _hipFireBobAmount * _weaponBobFactor;
+            var vBobValue = (Mathf.Sin(Time.time * _weaponBobFrequency * 2f) * 0.5f + 0.5f) * _weaponBobFactor * _hipFireBobAmount;
+            var xPosition = hBobValue;
+            var yPosition = Mathf.Abs(vBobValue);
+
+            _weaponBobLocalPosition = new Vector3(xPosition, yPosition, 0);
+        }
+        
+        private bool CanBobWeapon()
+        {
+            return (_characterMovement.IsWallRunning || _characterMovement.IsGrounded) 
+                   && !_characterMovement.IsDashing 
+                   && !_characterMovement.IsSliding;
+        }
+
+        private void UpdateWeaponRecoil()
+        {
+            if (_weaponRecoilLocalPosition.z >= _accumulatedRecoil.z * 0.99f)
```

Comment:
Probably should pull this out to a const, maybe the numbers in vBobValue too though those are less of a big deal imo

Reply:
yeah i don't actually know what these values are for, i got them from unity's tuts. Il'l pull them out and try to name it appropriately